Tilt switch

JerryWorkman opened the pull request:
**Description:**
Tilt switch for [Home Assistant](https://github.com/home-assistant/home-assistant)

Combines a tilt sensor and a relay switch to control a garage door opener or motorized gate.
When this switch is turned on it will toggle the relay to activate the garage door if it is
not already open. The reverse will happen when it is turned off.

configuration.yaml

switch tilt:
  platform: tilt
  switches:
    front_garage_door:
      tilt_sensor:   binary_sensor.my_tilt_switch
      switch:        switch.my_relay_switch
      contact_delay: 1  #optional on time for switch to simulate button press. default: 1 second
      run_time:      10 #optional run time for the opener. default: 10 seconds

JerryWorkman committed:
Add files via upload

JerryWorkman committed:
Delete tilt.py

JerryWorkman committed:
Add files via upload

Danielhiversen commented:
Thanks for your PR.

There are some failing tests here: https://travis-ci.org/home-assistant/home-assistant/jobs/184137932

When you submitted the PR, there was a checklist, that you are supposed to fill in:

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] Tests have been added to verify that the new code works.

JerryWorkman commented:
New to this... I'll be chipping away at the list in the next few days. I have several other components in the wings.

JerryWorkman committed:
Update with edits to satisfy PEP problems

JerryWorkman committed:
Update tilt.py

Danielhiversen commented on homeassistant/components/switch/tilt.py:
```diff
@@ -0,0 +1,154 @@
+    @property
+    def is_on(self):
+        """@:returns True if on"""
+        return self._is_on()
```
You should not  do I/O inside properties. More info here: https://github.com/home-assistant/home-assistant/issues/4210

Danielhiversen commented on homeassistant/components/switch/tilt.py:
```diff
@@ -0,0 +1,154 @@
+"""
+Tilt switch.
```
Change the comments here according to other components.

Danielhiversen reviewed (changes requested):
I am not sure I understand this component,
I just added some comments. There are probably more that needs to be fixed.

JerryWorkman committed:
Multiple edits to fix pylint issues and avoid I/O except in update()

arsaboo commented:
@Danielhiversen Many of us (including myself) use a tilt sensor and z-wave relay switch to automate our garage doors. It works very well, but requires additional automation (e.g., to reset the relay switch). It will be nice to have a built-in component like this :)

JerryWorkman committed:
Fix state reporting bug.

Added STATE_ON, STATE_OFF constants

JerryWorkman commented:
I'm an engineer who knows some programming. This is my first experience in
a major GitHub project. Any any all advice is appreciated.

I have also seen the relay switch "stick" closed which leaves the garage
door opener inoperable. Perhaps I should reset the switch at startup and at
the end of the second delay.

On Fri, Dec 16, 2016 at 3:51 PM, Alok Saboo <notifications@github.com>
wrote:

> @Danielhiversen <https://github.com/Danielhiversen> Many of us (including
> myself) use a tilt sensor and z-wave relay switch to automate our garage
> doors. It works very well, but requires additional automation (e.g., to
> reset the relay switch). It will be nice to have a built-in component like
> this :)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/4921#issuecomment-267692994>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABsbQHOpT35Au94V0NTrejQi2SB-rgupks5rIvnmgaJpZM4LNsHh>
> .
>

JerryWorkman committed:
Turn relay switch off at startup

 and after second (run_time) delay to insure that it is not stuck in the on position.

JerryWorkman committed:
Remove extranious comment

arsaboo commented:
@JerryWorkman See the automation that I am currently using to reset the relay after it is triggered [here](https://github.com/arsaboo/homeassistant-config/blob/master/automation.yaml#L535-L545). I am hoping that it will part of the component.

pvizeli commented on homeassistant/components/switch/tilt.py:
```diff
@@ -0,0 +1,131 @@
+        if self._running:
+            _LOGGER.debug('tilt switch %s running, assumed state is %s',
+                          self._name, self._state)
+            return self._state
```
update have no return value. Use only `return`

pvizeli commented on homeassistant/components/switch/tilt.py:
```diff
@@ -0,0 +1,131 @@
+            _LOGGER.debug('tilt switch %s is %s', self._name, tilt)
+
+    @property
+    def is_on(self):
```
That is done by base class and not allow to overwrite this on entity. Please remove

pvizeli commented on homeassistant/components/switch/tilt.py:
```diff
@@ -0,0 +1,131 @@
+        """@:returns True if on"""
+        return True if self._state == STATE_ON else False
+
+    def toggle(self):
```
please rename this to a internal function. ToggleEntity have a own toggle and they is not allow to overwrite here.

pvizeli commented on homeassistant/components/switch/tilt.py:
```diff
@@ -0,0 +1,131 @@
+            _LOGGER.debug('tilt opening %s', self._name)
+            self.toggle()
+        self._state = STATE_ON
+        self.update_ha_state()
```
use `self.schedule_update_ha_state()`

pvizeli commented on homeassistant/components/switch/tilt.py:
```diff
@@ -0,0 +1,131 @@
+            _LOGGER.debug('closing %s', self._name)
+            self.toggle()
+        self._state = STATE_OFF
+        self.update_ha_state()
```
use `self.schedule_update_ha_state()`

JerryWorkman replied:
Thanks. Testing all diffs now. Still working on the asyncio thing.

pvizeli reviewed (changes requested):
It is a fake device, you should they make async. The workflow don't allow to run in a executor and go back to async core. They will end in a deadlock after some time.

You need also add unittest for every functionality.
If that run in async context and with unittest, you become a approve from me.

A other question is, if you will make a generic_cover with this as cover platform like generic_thermostat or other generic stuff.

pvizeli commented on homeassistant/components/switch/tilt.py:
```diff
@@ -0,0 +1,131 @@
+        core.turn_on(self._hass, self._switch)
+        self._running = True
+        time.sleep(self._contact_delay)
+        core.turn_off(self._hass, self._switch)
```
Use a service call and not include the components.

pvizeli commented on homeassistant/components/switch/tilt.py:
```diff
@@ -0,0 +1,131 @@
+            self._insure_relay_is_off()
+            self._state = self._get_state()
+
+        core.turn_on(self._hass, self._switch)
```
Use a service call and not include the components.

JerryWorkman commented:
Can someone point me to a suitable example of a component that uses async
io?

On Fri, Dec 16, 2016 at 7:08 PM, Pascal Vizeli <notifications@github.com>
wrote:

> *@pvizeli* commented on this pull request.
> ------------------------------
>
> In homeassistant/components/switch/tilt.py
> <https://github.com/home-assistant/home-assistant/pull/4921#pullrequestreview-13432512>
> :
>
> > @@ -0,0 +1,131 @@
> +"""
> +Tilt Switch
> +custom component for Home Assistant https://home-assistant.io/
> +Documentation: https://home-assistant.io/components/switch.tilt/
> +Jerry Workman <jerry.workman@gmail.com>
> +"""
> +
> +import time
> +from threading import Timer
> +import logging
> +from homeassistant.components.switch import SwitchDevice
> +from homeassistant.const import (STATE_ON, STATE_OFF)
> +import homeassistant.components as core
>
> Don't import that. Use hass.services.call or hass.services.async_call
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/4921#pullrequestreview-13432512>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABsbQL3KNJ3o5dZ9xOxh2NVCON8sVsoPks5rIygagaJpZM4LNsHh>
> .
>

JerryWorkman committed:
Update tilt.py with recomendations

Removed core references and wait. Use timers for both delayed actions.
Still looking for asyncio and unittest examples.

JerryWorkman committed:
Update tilt.py

Make the hound happy

MartinHjelmare commented:
As @pvizeli mentioned I think you should make this a platform under the cover component.

JerryWorkman commented:
Making it a switch allows me to control it using automation. Can I use
automation to control a cover?

On Mon, Dec 19, 2016 at 9:41 PM, Martin Hjelmare <notifications@github.com>
wrote:

> As @pvizeli <https://github.com/pvizeli> mentioned I think you should
> make this a platform under the cover component.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/4921#issuecomment-268139042>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABsbQF0pf8GKAmtfKmMjgB4MesePV8Ffks5rJ0BpgaJpZM4LNsHh>
> .
>

JerryWorkman commented:
Looks like I can...

On Mon, Dec 19, 2016 at 9:45 PM, Jerry Workman <jerry.workman@gmail.com>
wrote:

> Making it a switch allows me to control it using automation. Can I use
> automation to control a cover?
>
> On Mon, Dec 19, 2016 at 9:41 PM, Martin Hjelmare <notifications@github.com
> > wrote:
>
>> As @pvizeli <https://github.com/pvizeli> mentioned I think you should
>> make this a platform under the cover component.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/home-assistant/home-assistant/pull/4921#issuecomment-268139042>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ABsbQF0pf8GKAmtfKmMjgB4MesePV8Ffks5rJ0BpgaJpZM4LNsHh>
>> .
>>
>
>

JerryWorkman commented:
Converting to cover.
